network_simulator/plot_tool.py

```diff
@@ -0,0 +1,300 @@
+        we would like to replace the previous version.
+        :param boolean scatter: if we want a scatter plot, we set this
+        variable to True; otherwise, it would be a line graph.
+        """
+        if not tuple_list:
+            return
+
+        if scatter:
+            # Scatter plot.
+            plt.scatter(*zip(*tuple_list), marker='o', linestyle='--')
+        else:
+            # Line graph
+            plt.plot(*zip(*tuple_list), marker='o', linestyle='--')
+
+    def graph_1d_list(self, lst=[]):
+        """
+        Graphs the content of a list.
+
+        :param list lst: list of values for the 1-D graph.
+        """
+        if not lst:
+            return
+
+        # Plot 1-D different values on a x-y plane.
+        plt.plot(lst, np.zeros_like(lst), marker='o', linestyle='--')
+
+class Grapher(object):
+    """
+    Graphs statistics items from Statistics class.
+    """
+    def __init__(self, stats=None):
+        """
+        :ivar Statistics stats: input Statistics object for graphing.
+        :ivar PlotTool tool: graphing functions.
+        """
+        self.stats = stats
+        self.tool = PlotTool()
+
+    def graph_links(self):
+        """
+        Graphs the link stats of each available link.
+        """
+        # Set up legend storage for link names.
+        occpy_legend = []
+        loss_legend = []
+        trans_legend = []
+
+        # Iterate through each link to display each stats.
+        for link, link_stats in self.stats.link_stats.iteritems():
+            # Name of link would be "link_name".
+            # Buffer occupancy w.r.t time.
+            plt.subplot(431)
+            self.tool.graph_tuple_list(link_stats.buffer_occupancy)
+            occpy_legend.append(link)
+
+            # Packet loss times.
+            plt.subplot(434)
+            loss_rate_lst = self.tool.gen_count_interval_list(
+                link_stats.packet_loss_times)
+            self.tool.graph_tuple_list(loss_rate_lst)
+            loss_legend.append(link)
+
+            # Packet transmission w.r.t time.
+            plt.subplot(437)
+            self.tool.graph_tuple_list(link_stats.packet_transmit_times)
+            trans_legend.append(link)
+
+        # Finalize buffer occupancy graph.
+        plt.subplot(431)
+        plt.legend(occpy_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Buffer Occupancy (pkts)")
+        # Set up grid on graph.
+        plt.grid(True)
+
+        # Finalize loss times graph.
+        plt.subplot(434)
+        plt.legend(loss_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Packet Losses")
+        # Set up grid on graph.
+        plt.grid(True)
+
+        # Finalize packet transmission graph.
+        plt.subplot(437)
+        plt.legend(trans_legend)
+        # Set up labels.
+        plt.xlabel("Time (sec)")
+        plt.ylabel("Size of Packet Transmitted (Bits)")
```

Comment:
Again, third graph for links should be the link flow rate in Mbps vs time. To put it in Mbps, you will have to take the rate tuple list and divide its rates by 1024^2.